### ISSUE TITLE
Implementability of hive::get_iterator(p)

### ISSUE THREAD

**Issue by Quuxplusone** [mattreecebentley/plf_hive#18]
```
int main()
{
    plf::hive<int> h = {1,2,3,4,5};
    auto it = std::next(h.begin(), 2);
    int *p = &(*it);
    assert(*p == 3);
    h.erase(std::next(h.begin(), 1), std::next(h.begin(), 4));
    auto pit = h.get_iterator(p);
    assert(pit == h.end());  // assert-fails
}
```

Here, I construct a hive `[1 2 3 4 5]`, take a pointer to element `3`, and then erase the middle of the hive, leaving `[1 _ _ _ 5]`. Then I call `h.get_iterator(p)`; it produces garbage instead of `h.end()`.

D0447R20 currently says (with <strike>one wording nit</strike>):

> Effects: Returns an `iterator` or `const_iterator` pointing to the same element as the <strike>`pointer` or</strike> `const_pointer`. If `p` does not point to an element in `*this`, the past-the-end iterator is returned.

Now, our `p` certainly doesn't point to an element in `*this` — but, it's also been invalidated by the `erase`. So is it supposed to be a valid argument for `get_iterator`? What kinds of pointers _are_ valid arguments for `get_iterator`? Could I pass a null pointer, for example?

`plf::hive`'s implementation of `get_iterator` relies on comparing pointers into arbitrary objects as if the memory address space on the target machine were flat, i.e. `p < q < r` implies that `q` is necessarily a member of the same object as `p` and `r`. _In general,_ C++ does not guarantee this: comparing pointers into arbitrary objects is UB, and comparing them with `std::less<void*>` is not UB but also isn't guaranteed to give you the answer you're looking for (because C++ allows pointer representations to be interleaved). Therefore, I believe `get_iterator` is unimplementable on a sufficiently hostile conforming implementation. Check with someone like @jwakely for absolute certainty, though.

(Oh yeah, and remember this all has to work for arbitrary user-defined fancy-pointer types too!)

Basically there are two prongs to this bug report:

(1) The `p < q < r` implementation approach involves UB; `get_iterator` is basically unimplementable in theory; I suggest `get_iterator` be struck from P0447 entirely.

(2) `plf::hive`'s own implementation of `get_iterator`, given a pointer to a previously valid element, will sometimes return a garbage iterator instead of `h.end()`; I call this a bug, or at least undesirable, and I wonder what you think about it.

**Comment by jwakely** [comment]
`std::less<T*>` should work, that's required to be a total order.

**Comment by Quuxplusone** [comment]
@jwakely: My understanding of the rules-as-written is that it works like this:
```
int a[2];
int b;
auto LT = std::less<int*>{};
assert(LT(&a[0], &a[1]));  // guaranteed
assert(LT(&a[0], &b) != LT(&b, &a[0]));  // one or the other must be true
assert(LT(&a[1], &b) != LT(&b, &a[1]));  // one or the other must be true
```
but that it is perfectly legal for an exotic implementation (maybe one that stores "segment" bits in the low-order bits of a pointer instead of the high-order bits) to produce an assertion failure for
```
assert(LT(&a[0], &b) == LT(&a[1], &b));  // NOT guaranteed, as I understand it
assert(LT(&a[0], &b) != LT(&b, &a[1]));  // NOT guaranteed, as I understand it
```
That is, as I understand it, the partial order imposed by `operator<` ensures that `a < a+1`, and the total order imposed by `LT` is consistent with the partial order, but nothing forbids an exotic implementation from claiming that `LT(a, &b) && LT(&b, a+1)`. (This is the question I was hoping you'd either agree with, or lay to rest.) On such an implementation, `plf::hive`'s pointer-comparison-based `get_iterator` would produce wrong answers.

On such an exotic implementation, iterating by incrementation from `&arr[0]` up to `&arr[9]` is still guaranteed to take 9 steps and touch all 10 elements _of `arr`;_ but that doesn't mean there aren't other unrelated objects hiding "between" `arr`'s objects from the point of view of `LT`'s total order.

**Comment by mattreecebentley** [comment]
On 29/04/2022 6:43 am, Quuxplusone wrote:
> (1) The |p < q < r| implementation approach involves UB; |get_iterator| 
> is basically unimplementable in theory; I suggest |get_iterator| be 
> struck from P0447 entirely.
> 
> (2) |plf::hive|'s own implementation of |get_iterator|, given a pointer 
> to a previously valid element, will sometimes return a garbage iterator 
> instead of |h.end()|; I call this a bug, or at least undesirable, and I 
> wonder what you think about it.
> 

I don't see how 1 makes sense and would still support C-style arrays but 
you can talk about it with Jonathan if you like. I don't think the 
standard library would be implementable in such an environment.
2 is a bug and thank you for finding it - it's a hangover from the HCJC 
skipfield which didn't allow zero middle nodes in skipblocks. The LCJC 
does, but this is easily fixed using the criteria in the LCJC paper 
under "7.2 Parallel Processing" and I have fixed this in the beta.
This info and the tech spec correction you've identified will be added 
to paper.

**Comment by opensdh** [comment]
The issue is that the implementation-defined total pointer order is not required to correspond with pointer arithmetic: there can exist _q_ such that _p_<_q_ (in that order, not C++'s `p<q`) and _q_<`p+1`.  There's nothing in the standard library that could detect such an (admittedly bizarre) implementation choice, because it never uses that total pointer order for anything that also cares about arrays.

**Comment by mattreecebentley** [comment]
Jonathan's said that std::less should work, so I'll just switch to using 
those.

**Comment by jwakely** [comment]
In practice on current implementations, yes, but not portably guaranteed by the standard, for the reasons given by Arthur and Davis.

Last time we tried to remove the allowance for weird, unordered address spaces there was strong opposition.

**Comment by mattreecebentley** [comment]
Closed as per LEWG discussion
